### PR TITLE
Identity provider aws iam auth e2e tests

### PR DESCRIPTION
**Pull request by pokearu**
*Issue #90*

*Description of changes:*
The changes include e2e tests for aws-iam-authenticator as an identity provider.

- Create cluster with aws-iam-authenticator enabled
- Test authentication using aws-iam-authenticator client to the cluster
- Delete cluster

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/556#" title="Author self-approved">pokearu</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [pokearu]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: internal/pkg/awsiam/download.go
Comment (abhay-krishna) [inline internal/pkg/awsiam/download.go:17]: nit: 
```suggestion
	k, err := getKernelName()
```
Same below.

File: internal/pkg/awsiam/download.go
Comment (abhay-krishna) [inline internal/pkg/awsiam/download.go:32]: Can probably use `runtime.GOOS`?

File: pkg/cluster/spec.go
Comment (abhay-krishna) [inline pkg/cluster/spec.go:516]:
```suggestion
	eksdRelease, err := s.getEksdRelease(versionsBundle)
```

File: test/framework/awsiam.go
Comment (abhay-krishna) [inline test/framework/awsiam.go:89]: Nit: For clarity
```suggestion
	return filepath.Join(e.ClusterName, fmt.Sprintf("%s-aws-iam-auth.kubeconfig", e.ClusterName))
```

Comment (pokearu) [reply test/framework/awsiam.go:89]: Oh actually that's the name we generate the kubeconfig in the CLI itself. Do you suggest we change that? [cluster_manager.go#L557](https://github.com/aws/eks-anywhere/blob/main/pkg/clustermanager/cluster_manager.go#L557)

File: scripts/e2e_test_docker.sh
Comment (pokearu) [reply scripts/e2e_test_docker.sh:56]: Change the regex in what way exactly? 
The gates flag just takes a list of feature gates to set to true. It does not impact the regex in anyway :)

File: cmd/integration_test/cmd/run.go
Comment (g-gaston) [inline cmd/integration_test/cmd/run.go:66]: Do we need this feature? Can't we just enable it in code as part of the "setup" process? It simplifies the interface of the cli and it's easier to find in code where at which flags are being enabled

File: internal/pkg/api/awsiam.go
Comment (g-gaston) [inline internal/pkg/api/awsiam.go:46]: nit: can we keep the "With" syntax for the options?

File: test/framework/awsiam.go
Comment (g-gaston) [inline test/framework/awsiam.go:19]: What's this for?

File: test/framework/awsiam.go
Comment (g-gaston) [inline test/framework/awsiam.go:46]: didn't we have something already to do this?

File: internal/pkg/awsiam/kubectl.go
Comment (g-gaston) [inline internal/pkg/awsiam/kubectl.go:18]: Don't need to change it, but couldn't we do this using the kubectl in `executables`?
we would just need to implement a new `Executable` that runs the magic to add `./bin` to the path
It seems more flexible and we don't duplicate code (plus the get pods method in `executables` already has unit tests

File: test/framework/awsiam.go
Comment (g-gaston) [inline test/framework/awsiam.go:67]: can we update this method, so it only downloads it if it doesn't exist?

File: internal/pkg/awsiam/kubectl.go
Comment (pokearu) [reply internal/pkg/awsiam/kubectl.go:18]: So the main reason i rewrote this was because the `kubectl` in executables uses the cli tools docker image.
When i am running kubectl here I pass in the kubeconfig for aws-iam-auth which has an `exec` to use the aws-iam-authenticator client binary which is downloaded in to `./bin`
When running kubectl through `docker run` the container does not have the aws-iam-auth client in it.
I am open to changing this, but I couldnt get an easy solution for using the downloaded local binary with `docker run`?

File: test/framework/awsiam.go
Comment (pokearu) [reply test/framework/awsiam.go:67]: I actually check file exists in the [download](https://github.com/aws/eks-anywhere/pull/556/files#diff-e06eba6891b95f2b1144ea8e44a270c424d4f541da0d1358c1be505844ac1989R18) method. Do you suggest to check here as well?

File: test/framework/awsiam.go
Comment (pokearu) [reply test/framework/awsiam.go:19]: So for creating a role mapping for iam-auth I get a RoleArn and a UserArn as an Env variable.
I am currently using only [RoleArn](https://github.com/aws/eks-anywhere/pull/556/files#diff-d9f3a86c63a87fd8ecc0a90afd5e7f4c91e6a71fb2c1103582dc3a2986d8f07aR41) but if someone wanted to test locally they could write a tests using their IAM user Arn.

File: cmd/integration_test/cmd/run.go
Comment (pokearu) [reply cmd/integration_test/cmd/run.go:66]: I was initially intending to do it in `setup`. But I thought then anyone who adds tests for a feature gated feature would need to add something in setup.
Instead my idea was we could enable all the gates we want for e2e testing from a command flag?

File: test/framework/awsiam.go
Comment (pokearu) [reply test/framework/awsiam.go:46]: I did check this and most of them return a particular "[option](https://github.com/aws/eks-anywhere/blob/main/internal/pkg/api/gitops.go#L72)" or [Type](https://github.com/aws/eks-anywhere/blob/main/internal/pkg/api/vsphere.go#L290). 
I just had this method for getting the value.
Is there something I am missing out?

File: cmd/integration_test/cmd/run.go
Comment (g-gaston) [reply cmd/integration_test/cmd/run.go:66]: The problem with that is the feature gate will be enabled even for the tests that "don't need it"

File: test/framework/awsiam.go
Comment (g-gaston) [reply test/framework/awsiam.go:19]: I see, but is this var being used anywhere?

File: internal/pkg/awsiam/kubectl.go
Comment (g-gaston) [reply internal/pkg/awsiam/kubectl.go:18]: The `kubectl` in `executables` doesn't always use the cli-tools image, it depends on the dependency we inject
For example, in the cli, it's possible to disable the cli-tools image and use directly the binaries from the host, it's just another type implementing the `Executable` interface
I'm proposing doing something similar, implementing that interface in a way that allows you to add the `iam-authenticator` to the path

File: cmd/integration_test/cmd/run.go
Comment (pokearu) [reply cmd/integration_test/cmd/run.go:66]: Actually I thought that might be helpful. Since it would ensure that all the tests are passing with the gate enabled.
So we know that the new feature doesn’t break any of the existing ones.

File: test/framework/awsiam.go
Comment (pokearu) [reply test/framework/awsiam.go:19]: `AWSIamUserArn` is currently not being used.
But it can be used similar to how `AWSIamRoleArn` while setting options for the test.
I could remove it if you feel it’s something that can be added later?

File: cmd/integration_test/cmd/run.go
Comment (pokearu) [reply cmd/integration_test/cmd/run.go]: As discussed, reverted changes in favor of enabling gates for corresponding tests.

File: test/framework/awsiam.go
Comment (pokearu) [reply test/framework/awsiam.go]: Removed const `AWSIamUserArn` to add if required later.

File: test/framework/awsiam.go
Comment (g-gaston) [inline test/framework/awsiam.go:94]: nit: no big deal but you could check first if this has already been added to the path before adding it
To prevent adding it twice

**Review by g-gaston — APPROVED** [review]
💯 🚀

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/556#pullrequestreview-804260608" title="Approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/556#" title="Author self-approved">pokearu</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston,pokearu]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
